File: client-flutter/lib/services/preference_service.dart
import 'api_client.dart';

/// Booking preferences — default trainer, training type, location.
/// Stored in backend `preference` table, cached locally per session.
class PreferenceService {
  Map<String, String?>? _cache;

  /// Load preferences from backend (or cache)
  Future<Map<String, String?>> getPreferences(String clientId) async {
    if (_cache != null) return Map.from(_cache!);

    final data = await apiClient.get('api/client/preferences/$clientId');
    if (data is Map<String, dynamic>) {
      _cache = {
        'trainer_id': data['trainer_id']?.toString(),
        'training_type_id': data['training_type_id']?.toString(),
        'location_id': data['location_id']?.toString(),
      };
      return Map.from(_cache!);
    }
    return {'trainer_id': null, 'training_type_id': null, 'location_id': null};
  }

  /// Save preferences to backend (partial update)
  Future<void> savePreferences(
    String clientId,
    Map<String, String?> prefs,
  ) async {
    final body = <String, dynamic>{};
    for (final key in ['trainer_id', 'training_type_id', 'location_id']) {
      if (prefs.containsKey(key)) {
        body[key] = prefs[key];
      }
    }
    if (body.isEmpty) return;

    await apiClient.put('api/client/preferences/$clientId', body: body);

    // Update cache
    _cache ??= {'trainer_id': null, 'training_type_id': null, 'location_id': null};
    for (final entry in prefs.entries) {
      _cache![entry.key] = entry.value;
    }
  }

  /// Clear cache (e.g. on logout)
  void clearCache() {
    _cache = null;
  }
}

File: client-ios/SihlClient/Services/PreferenceService.swift
import Foundation

/// Buchungs-Präferenzen (Standard-Trainer, Trainingstyp, Standort).
/// Pendant zu `services/preference_service.dart`.
/// Cached lokal; sync mit Backend bei Änderung.
actor PreferenceService {
    private var cache: [String: String?]?

    func getPreferences(clientId: String) async throws -> [String: String?] {
        if let cached = cache { return cached }

        let data = try await APIClient.shared.get("api/client/preferences/\(clientId)")
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return emptyPrefs()
        }
        let prefs: [String: String?] = [
            "trainer_id":       json["trainer_id"]       as? String,
            "training_type_id": json["training_type_id"] as? String,
            "location_id":      json["location_id"]      as? String,
        ]
        cache = prefs
        return prefs
    }

    func savePreferences(clientId: String, prefs: [String: String?]) async throws {
        var body: [String: Any] = [:]
        for key in ["trainer_id", "training_type_id", "location_id"] {
            if let val = prefs[key] {
                body[key] = val as Any
            }
        }
        guard !body.isEmpty else { return }
        _ = try await APIClient.shared.put("api/client/preferences/\(clientId)", body: body)
        cache = cache.map { c in
            var updated = c
            for (k, v) in prefs { updated[k] = v }
            return updated
        } ?? prefs
    }

    func clearCache() { cache = nil }

    private func emptyPrefs() -> [String: String?] {
        ["trainer_id": nil, "training_type_id": nil, "location_id": nil]
    }
}
